rclrs/src/parameter.rs

```diff
@@ -605,9 +454,101 @@ enum ParameterStorage {
+                    DeclaredValue::Mandatory(v) => v.read().unwrap().rcl_parameter_type(),
+                    DeclaredValue::Optional(v) => v
+                        .read()
+                        .unwrap()
+                        .as_ref()
+                        .map(|v| v.rcl_parameter_type())
+                        .unwrap_or(ParameterType::PARAMETER_NOT_SET),
+                    DeclaredValue::ReadOnly(v) => v.rcl_parameter_type(),
+                },
+            },
+            ParameterStorage::Undeclared(value) => value.rcl_parameter_type(),
+        }
+    }
+}
+
 #[derive(Debug, Default)]
-struct ParameterMap {
+pub(crate) struct ParameterMap {
     storage: BTreeMap<Arc<str>, ParameterStorage>,
+    allow_undeclared: bool,
+}
+
+impl ParameterMap {
+    /// Validates the requested parameter setting and returns an error if the requested value is
+    /// not valid.
+    fn validate_parameter_setting(
+        &self,
+        name: &str,
+        value: RmwParameterValue,
+    ) -> Result<ParameterValue, &str> {
+        let Ok(value): Result<ParameterValue, _> = value.try_into() else {
+            return Err("Invalid parameter type");
+        };
+        match self.storage.get(name) {
+            Some(entry) => {
+                if let ParameterStorage::Declared(storage) = entry {
+                    if std::mem::discriminant(&storage.kind)
+                        == std::mem::discriminant(&value.static_kind())
+                        || matches!(storage.kind, ParameterKind::Dynamic)
+                    {
+                        if !storage.options.ranges.in_range(&value) {
+                            return Err("Parameter value is out of range");
+                        }
+                        if matches!(&storage.value, DeclaredValue::ReadOnly(_)) {
+                            return Err("Parameter is read only");
+                        }
+                    } else {
+                        return Err(
+                            "Parameter set to different type and dynamic typing is disabled",
+                        );
+                    }
+                }
+            }
+            None => {
+                if !self.allow_undeclared {
+                    return Err(
+                        "Parameter was not declared and undeclared parameters are not allowed",
+                    );
+                }
+            }
+        }
+        Ok(value)
+    }
+
+    /// Stores the requested parameter in the map.
+    fn store_parameter(&mut self, name: Arc<str>, value: ParameterValue) {
+        match self.storage.entry(name) {
+            Entry::Occupied(mut entry) => match entry.get_mut() {
+                ParameterStorage::Declared(storage) => match &storage.value {
+                    DeclaredValue::Mandatory(p) => *p.write().unwrap() = value,
+                    DeclaredValue::Optional(p) => *p.write().unwrap() = Some(value),
+                    DeclaredValue::ReadOnly(_) => {}
```

Comment:
Yea I agree it should never happen but in case it does an `unreachable` panic is better than a silent failure, done https://github.com/ros2-rust/ros2_rust/pull/342/commits/a7d1befbe078418fca4852b200e3e1b6c4a62f9f